modules/shared/src/main/scala/io/constellationnetwork/security/Hash.scala

```diff
@@ -24,12 +24,53 @@ object hash {
   }
 
   object Hash {
+    private val hexDigits = "0123456789abcdef".toCharArray
+    private val sha256 = MessageDigest.getInstance("SHA-256")
 
+    trait Sha256Digest {
+      def toByteArray: Array[Byte]
+
+      def toHexString: String
+
+      private[constellationnetwork] def toIndexedSeq: IndexedSeq[Byte]
+    }
+
+    object Sha256Digest {
+      implicit val show: Show[Sha256Digest] = Show.show(s => s"Sha256Digest(${s.toHexString})")
+
+      def apply(bytes: Array[Byte]): Either[String, Sha256Digest] =
+        if (bytes.length == 32) {
+          Right(Sha256DigestImpl(bytes))
+        } else {
+          Left("invalid SHA-256 digest length")
+        }
+
+      private[constellationnetwork] case class Sha256DigestImpl(private val bytes: Array[Byte]) extends Sha256Digest {
+        override def toByteArray: Array[Byte] = Array.copyOf(bytes, bytes.length)
+
+        override def toHexString: String =
+          bytes
+            .foldLeft(new StringBuilder(bytes.length * 2)) { (sb, b) =>
+              sb.append(hexDigits((b >> 4) & 0xf)).append(hexDigits(b & 0xf))
+            }
+            .toString
+
+        override private[constellationnetwork] def toIndexedSeq: IndexedSeq[Byte] = bytes.toIndexedSeq
+      }
+    }
+
+    @deprecated("Use sha256DigestFromBytes() instead", since = "3.0")
     def hashCodeFromBytes(bytes: Array[Byte]): HashCode =
       Hashing.sha256().hashBytes(bytes)
 
+    def sha256DigestFromBytes(bytes: Array[Byte]): Sha256Digest = {
+      val md = sha256.clone().asInstanceOf[MessageDigest]
+      md.update(bytes)
+      Sha256Digest.Sha256DigestImpl(md.digest())
```

Comment:
hmm, you have created safe a apply function, but you don't use it at all 🤔

Reply:
I don't need it here. I know that `MessageDigest.digest()` will always return exactly 32 bytes, and if it ever doesn't, that would indicate a catastrophic failure. In that case, the number of bytes returned by `MessageDigest.digest()` would be the least of my concerns.